### PR TITLE
Check DLQ topic permissions on create/alter with MLP MOVE policy

### PR DESCRIPTION
**Pull request by nshestakov**
### Changelog entry 

Check write/alter permissions on a Dead Letter Queue topic when creating or altering a topic with an MLP consumer that uses MOVE DLQ.

### Description for reviewers 

При create/alter топика с MLP-консьюмером и DLQ (`MOVE`) проверяем, что у пользователя есть право **писать** (`UpdateRow`) **или альтерить** (`AlterSchema`) целевой DLQ-топик.

#### Зачем
Раньше можно было указать чужой/недоступный DLQ-топик. Сообщения туда уходили бы от имени системы, минуя права пользователя на запись.

#### Как работает
- Enforcement в `tx_proxy` (`schemereq.cpp`): для `Create/AlterPersQueueGroup` дополнительно resolve DLQ-путей и ACL any-of (`AlterSchema` | `UpdateRow`).
- На alter проверяются только **новые / изменившиеся** DLQ (не все существующие), чтобы unrelated alter не требовал прав на уже настроенный DLQ.
- Для этого в конфиг добавлены:
  - `TopicConfigVersion` — инкремент на alter
  - `ModificationVersion` у consumer — обновляется при add/change DLQ
- Helper'ы: `GetNormalizedDLQTopicPath` / `CollectDLQTopicPaths`, `CanonizeAndNormalizePath`.
- Internal-операции (`SetInternal(true)`) проверку DLQ пропускают (CDC, import, autoscaler, …).
- SQS-style DLQ (`sqs://…`) не проверяются как topic path.

#### Покрытие API
- Topic API / PQv1 create & alter (включая add/remove consumer)
- Версионирование consumers в Topic API и PQv1 alter paths

#### Tests
- `ydb/core/persqueue/public/schema/ut/dlq_acl_ut.cpp`
- `ydb/services/persqueue_v1/actors/schema/pqv1/dlq_acl_ut.cpp`
- правки связанных SDK / YQL / SQS / MLP consumer tests

Большая часть diff — тесты; прод-логика в основном в `schemereq`, schema versioning и `config`.

### CODE REVIEW

File: ydb/core/persqueue/public/schema/common.cpp [hunk 1]
```diff
@@ -412,7 +412,24 @@ TResult AddConsumer(
         consumersAdvancedMonitoringSettings->UpdateConsumerConfig(consumerConfig.name(), *consumer);
     }
 
+    UpdateConsumerVersion(*consumer, *config);
+
     return TResult();
 }
 
+void InitTopicConfigVersion(NKikimrPQ::TPQTabletConfig& config) {
```
Comment (kuzin57) [inline ydb/core/persqueue/public/schema/common.cpp]: это же протобуф, зачем 0 ставить форсированно?

File: ydb/public/lib/ydb_cli/dump/restore_impl.cpp [hunk 2]
```diff
@@ -1079,13 +1079,28 @@ TRestoreResult TRestoreClient::RestoreFolder(
...
+
+    // DLQ and other standalone topics must exist before changefeed consumers are restored.
+    for (const auto& entry : backupEntries) {
+        if (entry.Type == ESchemeEntryType::Topic) {
+            if (auto result = restoreEntry(entry); !result.IsSuccess()) {
+                return result;
+            }
+        }
+    }
+    for (const auto& entry : backupEntries) {
```
Comment (kuzin57) [inline ydb/public/lib/ydb_cli/dump/restore_impl.cpp]: зачем эти два цикла подряд?

**Review by kuzin57 — DISMISSED** [review]

File: ydb/library/aclib/aclib.cpp [hunk 3]
```diff
@@ -205,6 +205,18 @@ bool TSecurityObject::CheckAccess(ui32 access, const TUserToken& user) const {
...
 
+bool TSecurityObject::CheckAnyAccess(ui32 access, const TUserToken& user) const {
+    if (user.IsSystemUser()) {
+        return true; // the system always has access
+    }
+    if (HasOwnerSID() && user.IsExist(GetOwnerSID())) {
+        return true; // the owner always has access
+    }
+
+    const ui32 effectiveAccess = GetEffectiveAccessRights(user);
```
Comment (yurikiselev) [inline ydb/library/aclib/aclib.cpp]: Надо подумать над более аккуратной проверкой. Сейчас очень просто начать использовать этот метод неверно. Есть не гранулярный грант USE например. Если его сюда передать, то будет ОК, даже если есть доступ только на connect. Или если хотят `ydb.generic.read` или что-то ещё, этот рид содержит в себе в том числе `ydb.granular.describe_schema` и этого гранта хватает, чтобы начать читать.

Как будто нужно принимать прямо набор масок и смотреть, что любая **полная** из них есть.

**Review by Enjection — CHANGES_REQUESTED** [review]
1. It looks like we do not check if the target DLQ is the right thing and not some cdc stream impl or anything else like this. Would be great to have at least test that covers this edge case.
2. Also is DLQ now required for CREATE TOPIC? Shouldn't user get proper error message if so?

File: ydb/core/persqueue/public/config.cpp [hunk 4]
```diff
@@ -75,6 +76,52 @@ const NKikimrPQ::TPQTabletConfig_TPartition* GetPartitionConfigFromAllPartitions
...
+        return {};
+    }
+
+    return dlq;
+}
+
+} // namespace
+
+TString GetNormalizedDLQTopicPath(const NKikimrPQ::TPQTabletConfig_TConsumer& consumer, const TString& database) {
+    return CanonizeAndNormalizePath(database, GetDLQTopicPath(consumer));
```
Comment (Enjection) [inline ydb/core/persqueue/public/config.cpp]: The ACL check resolves the DLQ against the client-supplied database heade, but in mlp_consumer.cpp resolves it at runtime as GetYdbDatabasePath() + "/" + dlq with no canonization. Possibly will break something

File: ydb/core/tx/tx_proxy/schemereq.cpp [hunk 5]
```diff
@@ -805,6 +820,30 @@ struct TBaseSchemeReq: public TActorBootstrapped<TDerived> {
...
+        const NKikimrPQ::TPQTabletConfig& config
+    ) {
+        if (pbModifyScheme.GetInternal()) {
+            return;
+        }
+
+        const auto dlqPaths = NPQ::CollectDLQTopicPaths(
+            config,
+            GetRequestProto().GetDatabaseName(),
+            config.GetTopicConfigVersion()
```
Comment (Enjection) [inline ydb/core/tx/tx_proxy/schemereq.cpp]: And one AI note on top, not sure should we take it seriously or not:
```markdown
This filter assumes every producer of AlterPersQueueGroup maintains the version pair, but only the 5 Topic-API/PQv1 alter strategies call BumpTopicConfigVersion. TUpdateSchemeActorBase::FillModifyScheme (pq_schema_actor.h:465) re-proposes the stored config verbatim, so for a topic created but never altered via the Topic API (version 0, consumers 0) every Kafka AlterConfigs and DataStreams UpdateShardCount/SetWriteQuota matches all consumers and re-runs the DLQ check on consumers it never touched — AccessDenied on a pure retention change, or SCHEME_ERROR if the DLQ was since dropped.
```

**Comment by nshestakov** [comment]
Addressed the review comments:

1. **DLQ must be a topic.** After resolve we require `KindTopic` (and `PQGroupInfo`). A table or CDC stream (`KindCdcStream`, e.g. `/Root/table/feed`) is rejected with `SCHEME_ERROR` and message `Dead letter queue path must be a topic`. Tests: `CreateTopicWithMlpConsumerFailsWhenDlqIsATable`, `CreateTopicWithMlpConsumerFailsWhenDlqIsACdcStream`. CDC `streamImpl` is a real PQ topic (`KindTopic`), so it is still accepted as a destination; the changefeed path itself is not.

2. **DLQ is not required for CREATE TOPIC.** Only `MOVE` with an empty queue was previously stored silently. Create now matches alter: `Dead letter queue cannot be empty` (`BAD_REQUEST`). Test: `CreateSharedConsumerEmptyDlqRejected`. Missing DLQ path is still `SCHEME_ERROR`.

Also: runtime DLQ path resolution now uses `CanonizeAndNormalizePath`, and Kafka/DataStreams alters go through the same consumer version tracking so they do not re-check unchanged DLQs.

File: ydb/core/persqueue/public/config.cpp (hunk 4 above)
Comment (nshestakov) [reply ydb/core/persqueue/public/config.cpp]: Aligned runtime resolution with the ACL check: `mlp_consumer` now uses `CanonizeAndNormalizePath(Database, dlq)` (same helper as `GetNormalizedDLQTopicPath`). `sqs://` destinations are still returned as-is.

Relative names like `dlq` and absolute `/Root/dlq` both canonize to `/Root/dlq`. The previous `Database + "/" + dlq` concatenation broke on paths that already had a leading slash after canonization, or that did not start with `Database + "/"`.

File: ydb/core/tx/tx_proxy/schemereq.cpp (hunk 5 above)
Comment (nshestakov) [reply ydb/core/tx/tx_proxy/schemereq.cpp]: Took this seriously. `TUpdateSchemeActorBase::FillModifyScheme` (Kafka AlterConfigs, DataStreams UpdateShardCount/SetWriteQuota, etc.) now bumps `TopicConfigVersion` and runs `ApplyConsumerVersionUpdates` against the stored config, same as Topic API / PQv1 alter.

Unrelated Kafka/DataStreams alters no longer match existing DLQ consumers (their `ModificationVersion` is restored), so they will not re-run the DLQ ACL check or fail with AccessDenied / SCHEME_ERROR on a retention-only change.

File: ydb/core/tx/tx_proxy/schemereq.cpp [hunk 6]
```diff
@@ -1512,7 +1575,10 @@ struct TBaseSchemeReq: public TActorBootstrapped<TDerived> {
...
             }
 
-            if (!entry.SecurityObject->CheckAccess(access, *UserToken)) {
+            bool hasAccess = anyAccess ? entry.SecurityObject->CheckAnyAccess(access, *UserToken)
+                : entry.SecurityObject->CheckAccess(access, *UserToken);
+
+            if (!hasAccess) {
                 const auto errString = MakeAccessDeniedError(ctx, entry.Path, TStringBuilder()
                     << "with access " << NACLib::AccessRightsToString(access)
                 );
```
[the remark this answers is not in the document]
Comment (nshestakov) [reply ydb/core/tx/tx_proxy/schemereq.cpp]: Done. The DLQ ACL error now says `with any of access rights AlterSchema or UpdateRow`, so it is clear that either right is enough.

File: ydb/core/persqueue/public/schema/alter_topic.cpp [hunk 7]
```diff
@@ -330,6 +338,13 @@ TResult ApplyChangesInt(
             return result;
         }
     }
+    for (auto& consumer : *pqTabletConfig->MutableConsumers()) {
+        if (const auto it = addedAtTopicConfigVersionByConsumer.find(consumer.GetName()); it != addedAtTopicConfigVersionByConsumer.end()) {
+            consumer.SetAddedAtTopicConfigVersion(it->second);
+        } else {
```
[the remark this answers is not in the document]
Comment (nshestakov) [reply ydb/core/persqueue/public/schema/alter_topic.cpp]: Fixed. `ApplyConsumerVersionUpdates` bumps `ModificationVersion` when the normalized DLQ path changes, so altering an existing consumer's DLQ re-runs the ACL check on the new topic. Covered by `AlterTopicMlpConsumerFailsWhenChangingDlqToTopicWithoutAccess`.

File: ydb/services/persqueue_v1/actors/schema/pqv1/alter_topic.cpp [hunk 8]
```diff
@@ -56,6 +64,15 @@ struct TAlterTopicStrategy: public NPQ::NSchema::IAlterTopicStrategy {
         }
 
         auto& targetTabletConfig = *targetConfig.MutablePQTabletConfig();
+        targetTabletConfig.SetTopicConfigVersion(sourceTabletConfig.GetTopicConfigVersion() + 1);
+        for (auto& consumer : *targetTabletConfig.MutableConsumers()) {
```
[the remark this answers is not in the document]
Comment (nshestakov) [reply ydb/services/persqueue_v1/actors/schema/pqv1/alter_topic.cpp]: Same fix on the PQv1 path: it also calls `ApplyConsumerVersionUpdates`. PQv1 tests include `AlterTopicMlpConsumerFailsWhenChangingDlqToTopicWithoutAccess`.

File: ydb/core/persqueue/public/schema/ut/dlq_acl_ut.cpp [hunk 9]
```diff
@@ -0,0 +1,494 @@
...
+    auto result = CreateAndAlterTopicWithDLQ(setup, userSid, MAIN_TOPIC, DLQ_TOPIC);
+
+    UNIT_ASSERT(result);
+    UNIT_ASSERT_VALUES_EQUAL_C(result->Status, Ydb::StatusIds::UNAUTHORIZED, result->ErrorMessage);
+    UNIT_ASSERT(!result->ErrorMessage.empty());
+}
+
+} // Y_UNIT_TEST_SUITE(DlqAcl)
+
+} // namespace NKikimr::NPQ::NSchema
```
[the remark this answers is not in the document]
Comment (nshestakov) [reply ydb/core/persqueue/public/schema/ut/dlq_acl_ut.cpp:691]: Added:
- changing an existing consumer's DLQ to a topic without access (Topic API and PQv1)
- a dedicated PQv1 suite in `ydb/services/persqueue_v1/actors/schema/pqv1/dlq_acl_ut.cpp`

`sqs://` destinations are skipped in `GetDLQTopicPath` and never go through scheme ACL resolve.

File: ydb/core/persqueue/public/schema/common.cpp [hunk 10]
```diff
@@ -412,7 +412,24 @@ TResult AddConsumer(
         consumersAdvancedMonitoringSettings->UpdateConsumerConfig(consumerConfig.name(), *consumer);
     }
 
+    MarkConsumerAddedAtCurrentTopicConfigVersion(*consumer, *config);
```
[the remark this answers is not in the document]
Comment (nshestakov) [reply ydb/core/persqueue/public/schema/common.cpp]: Acknowledged. The old `AddedAtTopicConfigVersion` marking is gone; versions are handled only by `ApplyConsumerVersionUpdates`.

File: ydb/services/persqueue_v1/actors/schema/pqv1/alter_topic.cpp [hunk 11]
```diff
@@ -56,6 +64,15 @@ struct TAlterTopicStrategy: public NPQ::NSchema::IAlterTopicStrategy {
         }
 
         auto& targetTabletConfig = *targetConfig.MutablePQTabletConfig();
+        targetTabletConfig.SetTopicConfigVersion(sourceTabletConfig.GetTopicConfigVersion() + 1);
```
[the remark this answers is not in the document]
Comment (nshestakov) [reply ydb/services/persqueue_v1/actors/schema/pqv1/alter_topic.cpp]: Extracted. Topic API and PQv1 both call `BumpTopicConfigVersion` + `ApplyConsumerVersionUpdates` from `common.cpp`.

File: ydb/services/persqueue_v1/actors/schema/pqv1/alter_topic.cpp [hunk 12]
```diff
@@ -35,6 +38,15 @@ struct TAlterTopicStrategy: public NPQ::NSchema::IAlterTopicStrategy {
...
         const auto& pqConfig = AppData()->PQConfig;
         const auto& sourceTabletConfig = sourceConfig.GetPQTabletConfig();
+
+        absl::flat_hash_map<TString, std::pair<ui64, TString>> oldConsumerInfoByName;
+        for (const auto& c : sourceTabletConfig.GetConsumers()) {
+            oldConsumerInfoByName[c.GetName()] = {
+                c.GetModificationVersion(),
+                NPQ::GetDLQTopicPath(c),
+            };
+        }
```
[the remark this answers is not in the document]
Comment (nshestakov) [reply ydb/services/persqueue_v1/actors/schema/pqv1/alter_topic.cpp]: Same as above — shared helpers in `common.h` / `common.cpp`.

File: ydb/core/persqueue/public/config.cpp [hunk 13]
```diff
@@ -75,6 +76,46 @@ const NKikimrPQ::TPQTabletConfig_TPartition* GetPartitionConfigFromAllPartitions
...
+        if (modificationVersion && consumer.GetModificationVersion() != *modificationVersion) {
+            continue;
+        }
+
+        const auto dlq = GetDLQTopicPath(consumer);
+        if (dlq.empty()) {
+            continue;
+        }
+
+        result.insert(NKikimr::NormalizePath(NKikimr::CanonizePath(database), NKikimr::CanonizePath(dlq)));
```
[the remark this answers is not in the document]
Comment (nshestakov) [reply ydb/core/persqueue/public/config.cpp]: Path resolution now uses `CanonizeAndNormalizePath`. Relative names and in-database absolute paths canonize correctly. Cross-database DLQ is not supported; such a path fails resolve. `sqs://` is left unchanged.

File: ydb/core/http_proxy/ut/sqs_topic_ut.cpp [hunk 14]
```diff
@@ -128,6 +128,11 @@ namespace {
 
     }
 
+    bool CreateDlqTopic(NYdb::TDriver& driver, const TString& dlqTopicName = "DeadLetterQueue") {
```
[the remark this answers is not in the document]
Comment (nshestakov) [reply ydb/core/http_proxy/ut/sqs_topic_ut.cpp:150]: Leaving the helpers local to each test suite. Signatures differ enough that a shared util would not pay off in this PR.

File: ydb/core/persqueue/public/schema/common.cpp (hunk 1 above)
Comment (nshestakov) [reply ydb/core/persqueue/public/schema/common.cpp]: Уже неактуально: `ModificationVersion` больше не форсируется в 0. Версия берётся из `TopicConfigVersion` только если DLQ реально изменился (`UpdateConsumerVersion` / restore previous version).

File: ydb/public/lib/ydb_cli/dump/restore_impl.cpp (hunk 2 above)
Comment (nshestakov) [reply ydb/public/lib/ydb_cli/dump/restore_impl.cpp]: Уже неактуально для текущего diff: `restore_impl.cpp` в PR больше нет, логика версий собрана в `ApplyConsumerVersionUpdates`.

File: ydb/core/tx/tx_proxy/schemereq.cpp [hunk 15]
```diff
@@ -1506,15 +1566,33 @@ struct TBaseSchemeReq: public TActorBootstrapped<TDerived> {
                 }
             }
 
-            if (allowACLBypass || access == NACLib::EAccessRights::NoAccess || !entry.SecurityObject) {
+            const bool skipAccessCheck = requestIt->RequireAnyOfAccess.empty()
+                && access == NACLib::EAccessRights::NoAccess;
+
+            if (allowACLBypass || skipAccessCheck || !entry.SecurityObject) {
```
[the remark this answers is not in the document]
Comment (nshestakov) [reply ydb/core/tx/tx_proxy/schemereq.cpp:1729]: Pre-existing fail-open for all `ResolveForACL` entries, not DLQ-specific. Topic objects always have a `SecurityObject` in practice. Changing this globally would be a separate behavior change.

File: ydb/core/tx/tx_proxy/schemereq.cpp [hunk 16]
```diff
@@ -1506,6 +1566,14 @@ struct TBaseSchemeReq: public TActorBootstrapped<TDerived> {
                 }
             }
 
+            // RequireAccess and RequireAnyOfAccess are mutually exclusive: when
+            // RequireAnyOfAccess is set the RequireAccess bitmask is ignored below.
+            Y_DEBUG_ABORT_UNLESS(
+                requestIt->RequireAnyOfAccess.empty()
```
[the remark this answers is not in the document]
Comment (nshestakov) [reply ydb/core/tx/tx_proxy/schemereq.cpp]: Keeping `Y_DEBUG_ABORT_UNLESS`. The two fields are never set together today; a release abort here would turn a programming mistake into an availability incident. The check documents the invariant for debug builds.

File: ydb/core/tx/tx_proxy/schemereq.cpp [hunk 17]
```diff
@@ -1515,15 +1575,39 @@ struct TBaseSchemeReq: public TActorBootstrapped<TDerived> {
...
+                && access == NACLib::EAccessRights::NoAccess;
+
+            if (allowACLBypass || skipAccessCheck || !entry.SecurityObject) {
                 ++resolveIt;
                 ++requestIt;
                 continue;
             }
 
-            if (!entry.SecurityObject->CheckAccess(access, *UserToken)) {
+            Y_DEBUG_ABORT_UNLESS(
```
[the remark this answers is not in the document]
Comment (nshestakov) [reply ydb/core/tx/tx_proxy/schemereq.cpp:1735]: Same as the other thread: keeping `Y_DEBUG_ABORT_UNLESS` to document the invariant without aborting release builds.

**Comment by nshestakov** [comment]
@Enjection addressed the review:

1. DLQ target must be a real topic. CDC streams and CDC `streamImpl` are rejected with `CDC stream cannot be used as a dead letter queue: <path>`. Tables get `Dead letter queue path must be a topic, got <path>`. Covered by actor tests and YQL tests.

2. MOVE without a queue is rejected with `Dead letter queue cannot be empty`. Missing DLQ path and missing DLQ permissions now surface their issue text in YQL (`Path ... does not exist`, `Access denied ... with any of access rights AlterSchema or UpdateRow`).

**Review by Enjection — DISMISSED** [review]

File: ydb/core/tx/tx_proxy/schemereq.cpp [hunk 18]
```diff
@@ -1574,14 +1720,30 @@ struct TBaseSchemeReq: public TActorBootstrapped<TDerived> {
...
+
+            bool hasAccess = requestIt->RequireAnyOfAccess.empty()
+                ? entry.SecurityObject->CheckAccess(access, *UserToken)
+                : CheckAnyOfAccess(*entry.SecurityObject, requestIt->RequireAnyOfAccess, *UserToken);
+
+            if (!hasAccess) {
+                const TString errString = requestIt->RequireAnyOfAccess.empty()
+                    ? MakeAccessDeniedError(ctx, entry.Path, access)
+                    : TStringBuilder()
+                        << MakeAccessDeniedError(ctx, entry.Path, "with any of access rights AlterSchema or UpdateRow")
```
[the remark this answers is not in the document]
Comment (nshestakov) [reply ydb/core/tx/tx_proxy/schemereq.cpp]: Done. The user-facing string now uses the no-part `MakeAccessDeniedError(ctx, entry.Path)` overload and appends `with any of access rights AlterSchema or UpdateRow` once.

File: ydb/core/tx/tx_proxy/schemereq.cpp [hunk 19]
```diff
@@ -1405,6 +1520,37 @@ struct TBaseSchemeReq: public TActorBootstrapped<TDerived> {
...
+            {
+                continue;
+            }
+
+            const TString path = CanonizePath(entry.Path);
+            const bool isCdc = entry.Kind == NSchemeCache::TSchemeCacheNavigate::KindCdcStream
+                || (entry.Kind == NSchemeCache::TSchemeCacheNavigate::KindTopic
+                    && entry.Self
+                    && entry.Self->Info.GetPathSubType() == NKikimrSchemeOp::EPathSubTypeStreamImpl);
+            if (!isCdc && entry.Kind == NSchemeCache::TSchemeCacheNavigate::KindTopic && entry.PQGroupInfo) {
```
[the remark this answers is not in the document]
Comment (nshestakov) [reply ydb/core/tx/tx_proxy/schemereq.cpp:1540]: Added a comment: `KindTopic` without `PQGroupInfo` means the topic has not finished resolving (e.g. still being created), so we reject it instead of treating it as a valid DLQ destination.

File: ydb/core/persqueue/public/config.cpp [hunk 20]
```diff
@@ -75,6 +76,51 @@ const NKikimrPQ::TPQTabletConfig_TPartition* GetPartitionConfigFromAllPartitions
...
+    if (dlq.empty() || dlq.StartsWith("sqs://"sv)) {
+        return {};
+    }
+
+    return dlq;
+}
+
+} // namespace
+
+TString GetNormalizedDLQTopicPath(const NKikimrPQ::TPQTabletConfig_TConsumer& consumer, const TString& database) {
```
[the remark this answers is not in the document]
Comment (nshestakov) [reply ydb/core/persqueue/public/config.cpp:100]: Out of scope for this ACL check. A self-referential DLQ on create cannot resolve because the topic does not exist yet. Preventing a runtime loop if someone later points a consumer at its own topic belongs in MLP consumer / schema validation, not in the permission check.

File: ydb/core/persqueue/public/schema/alter_topic.cpp [hunk 21]
```diff
@@ -461,20 +467,22 @@ struct TAlterTopicStrategy: public IAlterTopicStrategy {
...
     }
 
+    const TString Database;
     Ydb::Topic::AlterTopicRequest Request;
 };
 
 } // namespace
 
 NActors::IActor* CreateAlterTopicActor(const NActors::TActorId& parentId, TAlterTopicSettings&& settings) {
+    const TString database = std::move(settings.Database);
```
[the remark this answers is not in the document]
Comment (nshestakov) [reply ydb/core/persqueue/public/schema/alter_topic.cpp]: Done. `CreateAlterTopicActor` now copies `settings.Database` directly into the operation and the strategy, without the intermediate move.
